torch/distributed/distributed_c10d.py

```diff
@@ -1156,6 +1157,148 @@ def all_gather_multigpu(output_tensor_lists,
         work.wait()
 
 
+def _object_to_tensor(obj):
+    buffer = io.BytesIO()
+    torch.save(obj, buffer)
+    input_tensor = torch.ByteTensor(list(buffer.getvalue()))
+    local_size = torch.LongTensor([input_tensor.numel()])
+    return input_tensor, local_size
+
+
+def _tensor_to_object(tensor, tensor_size):
+    buf = io.BytesIO(tensor.numpy().tobytes()[:tensor_size])
+    buf.seek(0)
+    out = torch.load(buf)
+    return out
+
+
+def all_gather_object(object_list, obj, group=group.WORLD):
+    """
+    Gathers picklable objects from the whole group into a list. Similar to
+    all_gather, but Python objects can be passed in. Note that the object must
+    be pickable in order to be gathered.
+
+    Arguments:
+        object_list (list[Any]): Output list. It should be correctly sized as the
+        size of the group for this collective and will contain the output.
+        object (Any): Pickable Python object to be broadcast from current process.
+        group (ProcessGroup, optional): The process group to work on
+
+    Returns:
+        None. If the calling rank is part of this group, the output of the
+        collective will be populated into the input object_list.
+
+    .. note:: Note that this API differs slightly from the all_gather collective
+        since it does not provide an async_op handle and thus will be a blocking
+        call.
+    """
+    if _rank_not_in_group(group):
+        return
+
+    input_tensor, local_size = _object_to_tensor(obj)
+    # Gather all local sizes. This is so that we can find the max size, and index
+    # until the correct size when deserializing the tensors.
+    group_size = get_world_size(group=group)
+    object_size_list = [torch.LongTensor([0]) for _ in range(group_size)]
+    gloo_group = None
+    # Avoid copying intermediate tensors back and forth to CUDA by using gloo PG
+    # for object collectives.
+    if get_backend(group) == Backend.NCCL:
+        gloo_group = new_group(
+            ranks=[i for i in range(get_world_size()) if not _rank_not_in_group(group)],
+            backend=Backend.GLOO,
+        )
+    # Allgather tensor sizes
+    all_gather(
+        object_size_list, local_size, group=group if gloo_group is None else gloo_group
+    )
+    max_object_size = max(object_size_list)
+    # allocate a tensor of max size, and copy byte tensor into it.
+    # TODO: Is there a more efficient way to do this / can we just resize inpt_tensor to max_object_size?
+    local_max_size_tensor = torch.ByteTensor(size=(max_object_size,))
+    local_max_size_tensor[: local_size.item()] = input_tensor
+    output_tensors = [
+        torch.empty(max_object_size, dtype=torch.uint8) for _ in range(group_size)
```

Comment:
(not sure if this can work) we can avoid multiple memory allocation by creating a big tensor and then let output tensors points to its views.

Reply:
We would still end up allocating the same amount of memory right? Although there may be a perf win by doing less memory allocation calls.

Reply:
Yep, exactly.